x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/Params.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.parser;
+
+import org.antlr.v4.runtime.Token;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class Params {
+
+    // This matches the named or unnamed parameters specified in an EsqlQueryRequest.params.
+    private List<Param> paramsByPosition = new ArrayList<>();
+
+    // This matches the named parameters specified in an EsqlQueryRequest.params.
+    private HashMap<String, Param> paramsByName = new HashMap<>();
+
+    // This is populated by EsqlParser, each parameter marker has an entry in paramsByToken
+    private Map<Token, Param> paramsByToken = new HashMap<>();
+
+    public Params() {}
+
+    public Params(List<Param> params) {
+        this.paramsByPosition.addAll(params);
+        populateNamedParams();
+    }
+
+    private void populateNamedParams() {
+        for (Param p : this.paramsByPosition) {
+            if (p.name != null) {
+                paramsByName.put(p.name, p);
+            }
+        }
+    }
+
+    public List<Param> positionalParams() {
+        return this.paramsByPosition;
+    }
+
+    public Param paramByPosition(int index) {
```

Comment:
Optional (since signature makes the intention obvious):  
```suggestion
    public Param get(int index) {
```